server/src/main/java/org/elasticsearch/cluster/routing/IndexRouting.java

```diff
@@ -129,8 +137,23 @@ public void checkIndexSplitAllowed() {}
 
         protected abstract int shardId(String id, @Nullable String routing);
 
+        @Override
+        public void process(IndexRequest indexRequest) {
+            if ("".equals(indexRequest.id())) {
+                throw new IllegalArgumentException("if _id is specified it must not be empty");
+            }
+
+            // generate id if not already provided
+            if (indexRequest.id() == null) {
+                indexRequest.autoGenerateId();
+            }
+        }
+
         @Override
         public int indexShard(String id, @Nullable String routing, XContentType sourceType, BytesReference source) {
+            if (id == null) {
+                throw new IllegalStateException("id is required and should have been set by process");
+            }
```

Comment:
This `null` test makes me think that maybe we should combine these two methods. We know the id can't be null because we just set it in `process`. But, like I said above, I kind of like that the shard routing methods don't mutate anything.